core/src/main/java/org/apache/calcite/tools/RelBuilder.java

```diff
@@ -1908,9 +1908,19 @@ private RelBuilder project_(
       fieldNameList.add(null);
     }
 
+    // Do not merge projections when top project contains RexSubQuery
+    boolean containsSubQuery = false;
+    for (RexNode node : nodes) {
+      if (RexUtil.containsSubQuery(node)) {
+        containsSubQuery = true;
+        break;
+      }
+    }
+
```

Comment:
I would move the check inside the if block (maybe after line 1934/1944) so that we don't spend time looking for subqueries if we don't need to.

Reply:
Also this is a breaking change since it changes the old behavior. It fixes a bug but at the same time prevents some optimizations to take effect. I think it is worth adding a few lines explaining the change in behavior in the respective section in `history.md`.